logstash-core/src/main/java/org/logstash/instrument/metrics/ExtendedFlowMetric.java

```diff
@@ -0,0 +1,360 @@
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *	http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.logstash.instrument.metrics;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.logstash.util.SetOnceReference;
+
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.VarHandle;
+import java.time.Duration;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.EnumSet;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.OptionalDouble;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.LongSupplier;
+import java.util.function.ToLongFunction;
+import java.util.stream.Collectors;
+
+import static org.logstash.instrument.metrics.FlowMetricRetentionPolicy.BuiltInRetentionPolicy;
+
+/**
+ * The {@link ExtendedFlowMetric} is an implementation of {@link FlowMetric} that produces
+ * a variable number of rates defined by {@link FlowMetricRetentionPolicy}-s, retaining no
+ * more {@link FlowCapture}s than necessary to satisfy the requested resolution or the
+ * requested retention. This implementation is lock-free and concurrency-safe.
+ */
+public class ExtendedFlowMetric extends BaseFlowMetric {
+    static final Logger LOGGER = LogManager.getLogger(ExtendedFlowMetric.class);
+
+    private final Collection<? extends FlowMetricRetentionPolicy> retentionPolicies;
+
+    // set-once atomic reference; see ExtendedFlowMetric#appendCapture(FlowCapture)
+    private final SetOnceReference<List<RetentionWindow>> retentionWindows = SetOnceReference.unset();
+
+    public ExtendedFlowMetric(final String name,
+                              final Metric<? extends Number> numeratorMetric,
+                              final Metric<? extends Number> denominatorMetric) {
+        this(System::nanoTime, name, numeratorMetric, denominatorMetric);
+    }
+
+    ExtendedFlowMetric(final LongSupplier nanoTimeSupplier,
+                       final String name,
+                       final Metric<? extends Number> numeratorMetric,
+                       final Metric<? extends Number> denominatorMetric,
+                       final Collection<? extends FlowMetricRetentionPolicy> retentionPolicies) {
+        super(nanoTimeSupplier, name, numeratorMetric, denominatorMetric);
+        this.retentionPolicies = List.copyOf(retentionPolicies);
```

Comment:
If our `lifetimeBaseline` was set by super, we should append it immediately (possibly creating our `RetentionWindow`s).

This might read more clearly if we implemented `SetOnceReference<T>#ifSet(Consumer<T>)`, but for now we can chain through `SetOnceReference<T>#asOptional()` to use `Optional<T>#ifPresent(Consumer<T>)`.

```suggestion
        this.retentionPolicies = List.copyOf(retentionPolicies);
        
        this.lifetimeBaseline.asOptional().ifPresent(this::appendCapture)
```